my-sms/spec/controllers/students_controller_spec.rb

```diff
@@ -23,119 +25,115 @@
-    it "returns a success response" do
-      student = Student.create! valid_attributes
-      get :edit, {:id => student.to_param}, valid_session
-      expect(response).to be_success
+  describe 'GET #edit' do
+    subject { get :edit, { id: student.to_param } }
+    it 'returns a success response' do
+      should be_success
+      should render_template('edit')
     end
   end
 
-  describe "POST #create" do
-    context "with valid params" do
-      it "creates a new Student" do
-        expect {
-          post :create, {:student => valid_attributes}, valid_session
-        }.to change(Student, :count).by(1)
+  describe 'POST #create' do
+    subject { post :create, { student: student } }
+
+    context 'with valid params' do
+      let(:student) { FactoryBot.attributes_for(:student) }
+
+      it 'creates a new Student' do
+        expect { subject }.to change(Student, :count).by(1)
       end
 
-      it "redirects to the created student" do
-        post :create, {:student => valid_attributes}, valid_session
-        expect(response).to redirect_to(Student.last)
+      it 'redirects to the created student' do
+        should be_redirect
+        should redirect_to(Student.last)
       end
     end
 
-    context "with invalid params" do
+    context 'with invalid params' do
+      let(:student) { FactoryBot.attributes_for(:student, first_name: nil) }
+
       it "returns a success response (i.e. to display the 'new' template)" do
-        post :create, {:student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
+        should render_template('new')
       end
     end
   end
 
-  describe "PUT #update" do
-    context "with valid params" do
-      let(:new_attributes) {
-        skip("Add a hash of attributes valid for your model")
-      }
+  describe 'PUT #update' do
+    let(:student) { FactoryBot.create(:student) }
+    subject { put :update, { id: student.to_param, student: new_attributes } }
+
+    context 'with valid params' do
+      let(:new_attributes) { FactoryBot.attributes_for(:student) }
 
-      it "updates the requested student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => new_attributes}, valid_session
+      it 'updates the requested student' do
+        subject
         student.reload
-        skip("Add assertions for updated state")
+        attrs = %i[first_name last_name gender]
+        student_attrs = student.attributes.fetch_values(*attrs.map(&:to_s))
+        new_attrs = new_attributes.fetch_values(*attrs)
+        expect(student_attrs).to eq new_attrs
       end
 
-      it "redirects to the student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => valid_attributes}, valid_session
-        expect(response).to redirect_to(student)
+      it 'redirects to the student' do
+        should be_redirect
```

Comment:
Probably don't need `should be_redirect` if you're following it with `should redirect_to(student)`, as the second assertion effectively encompasses the first. That being said, it also doesn't hurt to leave `should be_redirect` here, as it just checks the response code, so I'm not fussed either way.

Reply:
I took this as following the same pattern as testing for a success code before testing which template is rendered, but I guess the relationship is reciprocal with a redirect.
I'll leave it in for now if its up to me as I'd prefer to keep the same pattern for pattern's sake even if it doesn't test anything.

Reply:
Yeah, I'm good with that. Being in the same block means it's not actually running the test twice, so I think it's reasonable to leave it there.